rfcs/0157-more-granular-control.md

```diff
@@ -0,0 +1,241 @@
+
+In void (just as a reference), a package that ignores already upstream-dependency is built (in a chroot by default). This output folder can then act as a simple ad-hoc repository mirror with any http server (including `python3 -m http.server -d hostdir/binpkgs` on the host-machine or even github pages after a build from GitHub actions).
+
+
+# Detailed design
+[design]: #detailed-design
+
+While the current configuration does work for most people, and many are content with the existing solutions, a more granular control system would offer tremendous possibilities for customization and make the infrastructure flexible for future shifts.
+
+Instead of deprecating pulseaudio's `sound.*` in favor of `services.pipewire.*`, we should aim to build a flexible API that persists even when new standards emerge, reducing confusion caused by generic names like "sound" being deprecated.
+
+### Everything is opt-in!
+
+First and foremost, during image building (whether it's nixos-rebuild or vm/iso generation), the filesystem should be empty, with no packages or files.
+
+We can then assemble basic components by defining choices such as the bootloader, kernel, initramfs generator [1], libc, init, and then the choice of sound & de/wm.
+
+This approach also facilitates the use of NixOS-modules on non-NixOS nix installations.
+
+<sub>[1]: Recently [dracut merged](https://github.com/NixOS/nixpkgs/issues/77858), but there are no clear instructions on how to use that.</sub>
+
+### Emphasis on flake over shell script wrappers
+
+With the introduction of `nix run` and `nix build`, all current implementations should transition to using these first-party nix commands instead of wrappers like nixos-rebuild, nixos-generate, and home-manager.
+
+Operating everything under flake also opens up various opportunities, such as building the latest nixos iso image with or without plasma/gnome directly from the nixpkgs repository if its derivation is exposed in flake.nix (for example, `nix build nixpkgs#isoInstallerImageWithGnome`).
+
+### CLI improvements
+
+As explained earlier in the motivation section, the binary cache has no direct access, and using `nix-store --import {exported}` does not automatically pull missing dependencies from the binary cache in case the export does not include the whole closure.
+
+Allowing direct binary-cache access, such as querying information and pulling specific paths into the local `/nix/store`, would be a significant enhancement.
+
+### (optional) stdenv improvements
+
+Optionally, we could accompany this with a simpler, low-abstraction alternative to stdenv, or an opt-in C compiler setup.
+
+It is worth noting that `stdenv.mkDerivation` pulls gcc and autotools, but not all builds require these dependencies. Some simple shell scripts and builds in other languages redundantly pull these dependencies as part of the build, even when builtins.derivation provided by nix language could be sufficient.
```

Comment:
`stdenvNoCC` is what you looking for I guess?

Reply:
If that is the case, then most of the packages and script wrappers unknowingly just use the `stdenv`, for instance this [iso generation shell script](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-iso9660-image.nix#L48) has nothing to do with c compiler yet still use stdenv. And there are hundred of other packages that do this too...

I think the use of `stdenvNoCC` when possible can be effectively enforced (for most of non-C derivations) with this RFC? If that is ok?

Reply:
> I think the use of stdenvNoCC when possible can be effectively enforced (for most of non-C derivations) with this RFC. If that is ok?

If no c compiler is needed, then switching `stdenv` to `stdenvNoCC` is absolutely fine IMHO. This is probably not so much of an issue to discuss, but somethign to be documented more prominently and in the end somebody has to do the work (i.e. "fixing" up the packages).

Reply:
> And there are hundred of other packages that do this too...

This is just a matter of pull-requesting a treewide modification. No need for an RFC.

Also, and again, it adds four completely decoupled components to this RFC.

> I think the use of `stdenvNoCC` when possible can be effectively enforced (for most of non-C derivations) with this RFC. If that is ok?

It is already enforced when code is reviewed.

You can basically remove this section, it is 100% redundant.